feat: `dnf` module (#377)
blue-build/modules: modules/dnf/README.md

# **`dnf` Module**

The `dnf` module offers pseudo-declarative package and repository management using [`dnf5`](https://github.com/rpm-software-management/dnf).

## Features

This module is capable of:

- Repository Management

>   - Adding repo files via url or local files
>   - Removing repos by specifying the repo name
>   - Automatically cleaning up any repos added in the module
>   - Adding keys for repos via url or local files
>   - Enabling/disabling COPR repos
>   - Adding non-free repos like `rpmfusion` and `negativo17`

Reviewer: IMO COPR is the most important of these

```suggestion
  - Enabling/disabling COPR repos
  - Adding repo files via url or local files
  - Removing repos by specifying the repo name
  - Automatically cleaning up any repos added in the module
  - Adding keys for repos via url or local files
  - Adding non-free repos like `rpmfusion` and `negativo17`
```

> - Package Management
>   - Installing packages via url, local rpm files, or repo packaging
>   - Specifying repos from which to install packages
>   - Removing packages
>   - Replacing packages with ones from another repo

Reviewer: ```suggestion
- Package Management
  - Installing packages from RPM urls, local RPM files, or package repositories
  - Installing packages from a specific repository
  - Removing packages
  - Replacing installed packages with versions from another repository
```

- Optfix
  - Setup symlinks to `/opt/` to allow certain packages to install

### Add Repository Files

> - Add repos from any `https://` or `http://` URL
> - Any `.repo` files located in `./files/dnf/` of your image repo

Reviewer: Writing advice: make your bullet points symmetrical (have the same grammatical structure).

Oh and, let's include docs about `%OS_VERSION%` here too.

```suggestion
- Add repos from
  - any `https://` or `http://` URL
  - any `.repo` files located in `./files/dnf/` of your image repo
- If the OS version is included in the file name or URL, you can substitute it with the `%OS_VERSION%` magic string
  - The version is gathered from the `VERSION_ID` field of `/usr/lib/os-release`
```

```yaml
type: dnf
repos:
  files:
    - https://brave-browser-rpm-release.s3.brave.com/brave-browser.repo
    - custom-file.repo # file path for /files/dnf/custom-file.repo
```

> ### Add COPR Repositories
> ```yaml

Reviewer: ```suggestion
### Add COPR Repositories

- [COPR](https://copr.fedorainfracloud.org/) contains software repositories maintained by fellow Fedora users

```yaml
```

type: dnf
repos:
  copr:
    - atim/starship
    - trixieua/mutter-patched
```

### Disable/Enable Repositories
```yaml
type: dnf
repos:
  files:
    add:
      - repo1
      - repo2
    remove:
      - repo3
  copr:
    enable:
      - ryanabx/cosmic-epoch
    disable:
      - kylegospo/oversteer
```

> ### Add Repository Keys
> ```yaml
> type: dnf
> repos:
>   keys:
>     - https://example.com/repo-1.asc
>     - key2.asc
> ```

Reviewer: I'm not sure what `key2.asc` is in the original version? The code doesn't seem to be structured in such a way that they could be local files, but if it's not meant to be a local file I'm not sure what it is then...

````suggestion
### Add Repository Keys

- If the OS version is included in the file name or URL, you can substitute it with the `%OS_VERSION%` magic string
  - The version is gathered from the `VERSION_ID` field of `/usr/lib/os-release`

```yaml
type: dnf
repos:
  keys:
    - https://brave-browser-rpm-release.s3.brave.com/brave-core.asc
    - key-file.asc
```
````

Author: As far as I see in code, only URL format is supported for specifying `.asc` key?

Correct me if I'm wrong

Reviewer 2: Guess I should add that in

Reviewer 2: Ok, you can now specify keys like you can with repo files. The key name will be expanded to `${CONFIG_DIR}/dnf/`.

### Packages

> ```yaml
> type: dnf
> install:
>   packages:
>     - repo: repo-1
>       packages:
>         - repo-1-package-1
>         - repo-1-package-2
>     - package-3
> ```

Reviewer: Not sure about the necessity of the comment I added, but I think the re-ordering is useful.

````suggestion
```yaml
type: dnf
install:
  packages:
    - package-1
    # you also can set packages to be installed from a specific repository
    - repo: repo-1
      packages:
        - package-2
        - package-3
```
````

> ### Packages from URL or File
> ```yaml
> type: dnf
> install:
>   packages:
>     - https://example.com/package-%OS_VERSION%.rpm
>     - custom-file.rpm # file path for /files/dnf/custom-file.rpm
> ```

Reviewer: ````suggestion
### Packages from URL or File

- If the OS version is included in the file name or URL, you can substitute it with the `%OS_VERSION%` magic string
  - The version is gathered from the `VERSION_ID` field of `/usr/lib/os-release`

```yaml
type: dnf
install:
  packages:
    - https://example.com/package-%OS_VERSION%.rpm
    - custom-file.rpm # file path for /files/dnf/custom-file.rpm
```
````

> ### Install Packages from Specific Repositories
> ```yaml
> type: dnf
> install:
>   packages:
>     - repo: copr:copr.fedorainfracloud.org:custom-user:custom-repo
>       packages:
>         - package1
> ```

Reviewer: I think it's important that we state explicitly that you cannot just plop a COPR repo name or the .repo file name here, it should be the actual name of the repository.

````suggestion
### Install Packages from Specific Repositories

- Set `repo` to the name of the RPM repository, not the name or URL of the repo file

```yaml
type: dnf
install:
  packages:
    - repo: copr:copr.fedorainfracloud.org:custom-user:custom-repo
      packages:
        - package1
```
````

### Remove Packages
```yaml
type: dnf
remove:
  packages:
    - package1
    - package-2
```

> ### Define Packages Groups
> ```yaml
> type: dnf
> group-install:
>   packages:
>     - de-package-1
>     - wm-package-2
> ```
> 
> ### Remove Packages Groups
> ```yaml
> type: dnf
> group-remove:
>   packages:
>     - de-package-2
> ```

Reviewer: Title refinement + helpful info

````suggestion
### Install Package Groups

- See list of all package groups by running `dnf5 group list --hidden` on a live system

```yaml
type: dnf
group-install:
  packages:
    - de-package-1
    - wm-package-2
```

### Remove Package Groups
```yaml
type: dnf
group-remove:
  packages:
    - de-package-2
```
````

### Replace Packages
```yaml
type: dnf
replace:
  - from-repo: copr:copr.fedorainfracloud.org:custom-user:custom-repo
    packages:
      - package-1
  - from-repo: repo-1
    packages:
      - old: old-package-2
        new: new-package-2
```

### Optfix
- Optfix is a script used to work around problems with certain packages that install into `/opt/`
  - These issues are caused by Fedora Atomic storing `/opt/` at the location `/var/opt/` by default, while `/var/` is only writeable on a live system
  - The script works around these issues by moving the folder to `/usr/lib/opt/` and creating the proper symlinks at runtime
- Specify a list of folders inside `/opt/`

```yaml
type: dnf
optfix:
  - package1
  - package2

> ```

Reviewer: ````suggestion
```yaml
type: dnf
optfix:
  - brave.com
  - foldername
```
````

### Add COPR Repositories

* Specify a list of COPR repositories in the `copr` field

Example:
```yaml
type: dnf
repos:
  copr:
    - atim/starship
    - trixieua/mutter-patched
```

### Add Repository Files

* Specify a URL or file path in the `files` field to add repository files
* Use flags such as `cleanup` to customize repository management

Example:
```yaml
type: dnf
repos:
  files:
    - https://brave-browser-rpm-release.s3.brave.com/brave-browser.repo
    - custom-file.repo # file path for /files/dnf/custom-file.repo
```

### Disable/Enable Repositories

* Specify a list of repositories to disable or enable in the `disable` or `enable` field

Example:
```yaml
type: dnf
repos:
  disable:
    - repo1
    - repo2
  enable:
    - repo3
```

### Add Repository Keys

* Specify a list of repository keys in the `keys` field

Example:
```yaml
type: dnf
repos:
  keys:
    - key1
    - key2
```

## Package Installation

### Install Packages

* Specify packages to install in the `install.packages` field
* Use the `repo` parameter to specify a specific repository for installation
* Use the `%OS_VERSION%` variable to automatically determine the operating system version

> * Use flags such as `skip-unavailable`, `install-weak-deps`, `skip-broken` and `allow-erasing` to customize package installation

Reviewer: I think there should be a section in the README dedicated to explaining what all the available flags are and what they mean. Language such as `such as` should be avoided here to avoid ambiguity. I would rather format this as
```
* You can specify the following flags with this operation
 * ...
```

install:
  packages:
    - https://github.com/Eugeny/tabby/releases/download/v1.0.209/tabby-1.0.209-linux-x64.rpm
    - custom-file.rpm # file path for /files/dnf/custom-file.rpm
```

### Install Packages from Specific Repositories

* Specify a repository in the `repo` field to install packages from that repository
* Use the `%OS_VERSION%` variable to automatically determine the operating system version

Example:
```yaml
type: dnf
install:
  packages:
    - repo: copr:copr.fedorainfracloud.org:trixieua:mutter-patched
      packages:
        - mutter
```

## Package Removal

### Remove Packages

* Specify packages to remove in the `remove.packages` field
* Use flags such as `auto-remove` to customize package removal

Example:
```yaml
type: dnf
remove:
  packages:
    - firefox
    - firefox-langpacks
```

## Package Group Installation

### Define Packages Groups

* Specify a package group in the `group-install.packages` field
* Use flags such as `skip-unavailable`, `install-weak-deps`, `skip-broken` and `allow-erasing` to customize package installation

Example:
```yaml
type: dnf
group-install:
  packages:
    - cosmic-desktop
    - window-managers
```

## Package Group Removal

> ### Remove Packages Groups

Reviewer: Remove all of these sorts of double-titles from the whole document. This is not a sensible structure.

```
## Doing a Specific Thing

### Do The Thing
```

Reviewer: When there are multiple `###`-level titles below the top-level title, that is fine. Might only need minor rewording.

* Specify a package group in the `group-remove.packages` field

Example:
```yaml
type: dnf
group-remove:
  packages:
    - development-tools
```

## Package Replacement

### Replace Packages

* Specify a replacement package in the `replace.from-repo` field
* If new package for replacement is named differently, you can use `old/new` format as outlined below
* Use flags such as `skip-unavailable`, `install-weak-deps`, `skip-broken` and `allow-erasing` to customize package installation

Example:
```yaml
type: dnf
replace:
  - from-repo: copr:copr.fedorainfracloud.org:trixieua:mutter-patched
    packages:
      - mutter
  - from-repo: fedora
    packages:
      - old: OpenCL-ICD-Loader
        new: ocl-icd
```

## Optfix

> ### Fix Optfix
> 
> * Specify a list of packages to fix optfix issues in the `optfix` field
> 
> Example:
> ```yaml
> type: dnf
> optfix:
>   packages:
>     - package1
>     - package2
> ```

Reviewer: There is no such thing as an "optfix fix". Optfix is the fix.

```suggestion
* Optfix is a script used to work around problems with certain packages that install into `/opt/`
  * These issues are caused by Fedora Atomic storing `/opt/` at the location `/var/opt/` by default, while `/var/` is only writeable on a live system
  * The script works around these issues by moving the folder to `/usr/lib/opt/` and creating the proper symlinks at runtime
* Specify a list of folders inside `/opt/`

Example:
```yaml
type: dnf
optfix:
  packages:
    - foldername # ADD AN ACTUAL EXAMPLE FOLDER HERE
```
```

Excuse me for the potentially problematic formatting.

## Known issues

Replacing the kernel with `dnf` module is not done cleanly & some remaints of old kernel will be present.  

> Please use `rpm-ostree` module for this purpose until this `dnf` behavior is fixed.

Reviewer: Was this fixed already in some images? I think I remember hearing about this.

Author: It will be fixed in Fedora 42 (`quay.io/fedora-ostree-desktops` images)
